### PR TITLE
[MPoW] Update stake limit in a long range

### PR DESCRIPTION
**Pull request by MyronFanQiu**

### CODE REVIEW

File: cstrml/swork/src/lib.rs [hunk 1]
```diff
@@ -306,8 +315,40 @@ decl_module! {
...
+            // 1. identity_previous_key is none and workload is none => not started
+            // 2. identity_previous_key is some and workload is some => calculating
+            // 3. identity_previous_key is none and workload is some => calculation is done and will send workload to staking module
+            if ((now + (UPDATE_OFFSET as u32)) % (REPORT_SLOT as u32)).is_zero() && Self::workload().is_none() && Self::identity_previous_key().is_none()  {
+                let prefix = <Identities<T>>::prefix_hash();
+                IdentityPreviousKey::put(prefix);
+                <Workload<T>>::put((BTreeMap::<T::AccountId, u128>::new(), 0u128, 0u128, 0u128));
+            }
+            // If it's not timeout and not finished yet, continue updating process
+            if !(now % (REPORT_SLOT as u32)).is_zero() && Self::identity_previous_key().is_some() {
```
Comment (MyronFanQiu) [inline cstrml/swork/src/lib.rs:331]: This line maybe change

File: cstrml/staking/src/lib.rs [hunk 2]
```diff
@@ -2463,6 +2462,7 @@ impl<T: Config> swork::Works<T::AccountId> for Module<T> {
         } else {
             Self::update_stage_two_stake_limit(workload_map, total_workload, total_stake_limit);
         }
+        Self::deposit_event(RawEvent::UpdateStakeLimitSuccess);
```
Comment (zikunfan) [inline cstrml/staking/src/lib.rs:2465]: Considering give how many identities updated?

File: cstrml/swork/src/lib.rs [hunk 3]
```diff
@@ -306,8 +315,40 @@ decl_module! {
...
+                let prefix = <Identities<T>>::prefix_hash();
+                IdentityPreviousKey::put(prefix);
+                <Workload<T>>::put((BTreeMap::<T::AccountId, u128>::new(), 0u128, 0u128, 0u128));
+            }
+            // If it's not timeout and not finished yet, continue updating process
+            if !((now + END_OFFSET) % (REPORT_SLOT as u32)).is_zero() && Self::identity_previous_key().is_some() {
+                let previous_key = Self::identity_previous_key().unwrap();
+                // Update the workload map in one batch iter, might kill the IdentityPreviousKey
+                // which means updating process is finished.
+                Self::update_one_identities_batch(previous_key);
```
Comment (zikunfan) [inline cstrml/swork/src/lib.rs:335]: `patial_update_identities`

File: cstrml/swork/src/lib.rs [hunk 4]
```diff
@@ -306,8 +315,40 @@ decl_module! {
...
-            if (now % <T as frame_system::Config>::BlockNumber::from(REPORT_SLOT as u32)).is_zero()  {
-			    Self::update_identities();
+            let now = TryInto::<u32>::try_into(now).ok().unwrap();
+            // At REPORT_SLOT - UPDATE_OFFSET blocks, updating process would start
+            // IdentityPreviousKey is used as the switch as well
+            // There are three status
+            // 1. identity_previous_key is none and workload is none => not started
+            // 2. identity_previous_key is some and workload is some => calculating
+            // 3. identity_previous_key is none and workload is some => calculation is done and will send workload to staking module
+            if ((now + UPDATE_OFFSET)% (REPORT_SLOT as u32)).is_zero() && Self::workload().is_none() && Self::identity_previous_key().is_none()  {
```
Comment (zikunfan) [inline cstrml/swork/src/lib.rs]: extra space

**Review by zikunfan — CHANGES_REQUESTED** [review]

**Review by zikunfan — CHANGES_REQUESTED** [review]

**Review by zikunfan — APPROVED** [review]

**Pull request merged**